cardano-cli/src/Cardano/CLI/Shelley/Run/Transaction.hs

```diff
@@ -179,49 +183,47 @@ readShelleyCert :: CertificateFile -> ExceptT ShelleyTxCmdError IO Certificate
--- 'SigningKey's which have a Byron and a Shelley constructor.
-readSigningKeyFiles :: [SigningKeyFile] -> ExceptT ShelleyTxCmdError IO [SigningKey]
+
+data SomeSigningKey
+  = ApiPaymentSigningKey !(Api.SigningKey Api.PaymentKey)
+  | ApiGenesisUTxOSigningKey !(Api.SigningKey Api.GenesisUTxOKey)
+  | ApiStakePoolSigningKey !(Api.SigningKey Api.StakePoolKey)
+
+readSigningKeyFiles :: [SigningKeyFile] -> ExceptT ShelleyTxCmdError IO [SomeSigningKey]
 readSigningKeyFiles files =
   newExceptT $ do
     xs <- mapM readSigningKeyFile files
     case partitionEithers xs of
       -- TODO: Would be nice to also provide a filepath to the signing key
       -- resulting in the error.
-      (e:_, _) -> pure $ Left (ShelleyTxSignKeyError (ReadSigningKeyError e))
+      (e:_, _) -> pure $ Left (ShelleyTxReadFileError e)
       ([], ys) -> pure $ Right ys
 
-readSigningKeyFile :: SigningKeyFile -> IO (Either TextViewFileError SigningKey)
+readSigningKeyFile
+  :: SigningKeyFile
+  -> IO (Either (Api.FileError Api.TextEnvelopeError) SomeSigningKey)
 readSigningKeyFile (SigningKeyFile skfile) =
-      readTextViewEncodedFile decodeAddressSigningKey skfile
-
--- The goal here is to read either a Byron or Shelley address signing key or a
--- Genesis UTxO signing key. The TextView provides functions to read one of
--- several file types, however this does not currently mesh well with the
--- key reading functions from the API package.
---
--- The API package provides parseSigningKeyView but this takes the whole
--- ByteString textview, rather than the structured TextView, so we cannot
--- compose that with readTextViewEncodedFile. It provides the lower level
--- signingKeyFromCBOR which returns too big an error type to fit here, so
--- we have to fudge it with a partial conversion.
-decodeAddressSigningKey :: TextView -> Either TextViewError SigningKey
-decodeAddressSigningKey tView = do
-    isGenesisUTxOKey <- expectTextViewOfTypes fileTypes tView
-    if isGenesisUTxOKey
-      then decodeFromTextView (SigningKeyShelley <$> fromCBOR) tView
-      else first (\(ApiTextView tve) -> tve)
-                 (signingKeyFromCBOR (tvRawCBOR tView))
+    Api.readFileTextEnvelopeAnyOf fileTypes skfile
   where
     fileTypes =
-      [ ("SigningKeyShelley", False)
-      , ("SigningKeyByron",   False)
-      , (renderKeyType (KeyTypeSigning GenesisUTxOKey), True)
-      , (renderKeyType (KeyTypeSigning (OperatorKey StakePoolOperatorKey)), True)
+      [ Api.FromSomeType (Api.AsSigningKey Api.AsPaymentKey) ApiPaymentSigningKey
+      -- TODO: Byron signing key
+      , Api.FromSomeType (Api.AsSigningKey Api.AsGenesisUTxOKey) ApiGenesisUTxOSigningKey
+      , Api.FromSomeType (Api.AsSigningKey Api.AsStakePoolKey) ApiStakePoolSigningKey
       ]
 
+-- | Convert a 'SomeSigningKey' (which wraps a new-style typed API signing
+-- key) to a 'SigningKey' (old-style API signing key).
+toOldApiSigningKey :: SomeSigningKey -> SigningKey
```

Comment:
Again here, I'm not sure what the consensus on having these conversion functions everywhere or putting them into a module.